chain/migrate.go

```diff
@@ -0,0 +1,97 @@
+package chain
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"go.sia.tech/core/consensus"
+	"go.sia.tech/core/types"
+)
+
+// A MigrationLogger logs the progress of a database migration.
+type MigrationLogger interface {
+	Printf(format string, v ...any)
+	SetProgress(percentage float64)
+}
+
+// MigrateDB upgrades the database to the latest version.
+func MigrateDB(db DB, n *consensus.Network, l MigrationLogger) error {
+	if db.Bucket(bVersion) == nil {
+		return nil // nothing to migrate
+	}
+	dbs := &DBStore{
+		db: db,
+		n:  n,
+	}
+
+	version := dbs.bucket(bVersion).getRaw(bVersion)
+	if version == nil {
+		version = []byte{1}
+	}
+	switch version[0] {
+	case 1:
+		l.Printf("Migrating database from version 1 to 2")
+		l.Printf("Computing new element tree")
+		cs := n.GenesisState()
+		v1Blocks := min(dbs.getHeight(), n.HardforkV2.RequireHeight)
+		seen := make(map[[2]uint64]types.Hash256)
+		var tree [][2][]byte
+		flush := func() error {
+			sort.Slice(tree, func(i, j int) bool {
+				return bytes.Compare(tree[i][0], tree[j][0]) < 0
+			})
+			bucket := dbs.bucket(bTree)
+			for _, kv := range tree {
+				bucket.putRaw(kv[0], kv[1])
```

Comment:
Killing the migration mid-process is a non issue right?

Reply:
in this case, yeah; if you fail in the middle, you lose all progress and start over from 0. If we add a much more expensive migration in the future, we'd probably want to track its progress in a temporary DB bucket or something.